clients/drcachesim/tracer/syscall_pt_trace.cpp

```diff
@@ -0,0 +1,138 @@
+/* **********************************************************
+ * Copyright (c) 2022 Google, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* syscall_pt_trace.cpp: module for recording kernel PT traces for every syscall. */
+
+#include <cstring>
+
+#include "dr_api.h"
+#include "drpttracer.h"
+#include "syscall_pt_trace.h"
+
+#ifndef BUILD_PT_TRACER
+#    error "This module requires the drpttracer extension."
+#endif
+
+#define RING_BUFFER_SIZE_SHIFT 8
+#define PT_DATA_FILE_NAME_SUFFIX ".pt"
+#define PT_METADAT_FILE_NAME_SUFFIX ".meta"
+
+syscall_pt_trace_t::syscall_pt_trace_t()
+    : write_file_func_(nullptr)
+    , pttracer_handle_ {}
+    , recorded_syscall_num_(0)
+    , recording_sysnum_(-1)
+    , drcontext_(nullptr)
+    , pt_dir_name_ { '\0' }
```

Comment:
Why not ""?

Reply:
I tried using "" to initialize a string in the constructer. It seems DR will crash if I do this. Is this a limit in DR? Or is it an issue of DR? @derekbruening 
The two crash codes are shown in below:

(1) case 1: crash before print
 
``` c++
class syscall_pt_trace_t {
private:
    std::string pt_dir_name_;
};

syscall_pt_trace_t::syscall_pt_trace_t()
    :pt_dir_name_ ("")
{
    // crash before print
    dr_fprintf(stderr, "reach here");
}
```
(2)  case 2: crash after print

``` c++
class syscall_pt_trace_t {
private:
    std::string pt_dir_name_;
};
syscall_pt_trace_t::syscall_pt_trace_t()
{
    dr_fprintf(stderr, "reach here");
    // crash after print
    pt_dir_name_ = "";
}
```

Reply:
Please elaborate on "crash": what precisely happens, at what callstack.

`stderr` is an incorrect argument to dr_fprintf which takes file descriptors (like STDERR) not FILE pointers.  Maybe that is what is causing problems: but you'd expect it to just fail; but if you're relying on that to tell you when something happened it would not be reliable.

We do not want to call malloc during tracing (b/c we statically link drmemtrace into apps) so do not use std::string after init time.  Is this used after init time?  There is a check in a test which will fail if malloc is called.

Reply:
Sorry for giving you some confusion. 
The crash occurs when I add a new field in tracer.cpp's per_thread_t, which is an instance of the class containing std::string . During the event_thread_init callback, the code executes to the line where the per_thread_t instance will be initialized. The client will crash. The code is shown as below:
``` c++
class syscall_pt_trace_t {
private:
    std::string test_;
};
syscall_pt_trace_t::syscall_pt_trace_t()
: test_("")
{
}
per_thread_t {
    ...
    syscall_pt_trace_t syscall_pt_trace;
}

static void
event_thread_init(void *drcontext)
{
    per_thread_t *data = (per_thread_t *)dr_thread_alloc(drcontext, sizeof(per_thread_t));
    DR_ASSERT(data != NULL);
    *data = {};  // crash in the this line
    ...
}
```
If I remove the 'test_' in syscall_pt_trace_t it works fine. I know that the initialization of std::string invokes some allocation function. Do you think that's why the crash happened?

Reply:
A class does not need explicit initialization to its default value: that is done automatically, so this `test_("")` is superfluous.

I would not expect that line to crash: as the comment says, it should be properly handled wrt classes inside the struct.  As with any crash in our own C++ code, look at the callstack to understand what is going on: where exactly is the crash point?  I would expect an error in zeroing that struct to only cause a crash later when accessing the vtable of a class field or something.  A crash on the zeroing line itself implies a read-only page or something??  Again, look at the callstack: no need to speculate when we can get real data.